### ISSUE TITLE
REPL errors "randomly" after fake Pkg prompt

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#54754]
Just as an example:

![image](https://github.com/JuliaLang/julia/assets/1282691/cf346563-2698-4b11-94ae-f4635bf66ad0)

```
(Base) julia> Base.explicit_loadUnhandled Task ERROR: TaskFailedException

    nested task error: ArgumentError: IOBuffer not marked
    Stacktrace:
     [1] reset
       @ ./io.jl:1429 [inlined]
     [2] reset
       @ ./stream.jl:1489 [inlined]
     [3] peek(s::TTY, ::Type{Char})
       @ Base ./stream.jl:1496
     [4] peek(t::REPL.Terminals.TTYTerminal, ::Type{Char})
       @ REPL.Terminals ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/Terminals.jl:161
     [5] (::REPL.LineEdit.var"#281#283"{REPL.Terminals.TTYTerminal})()
       @ REPL.LineEdit ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:2834
Stacktrace:
 [1] _wait_multiple(waiting_tasks::Tuple{Task, Task}, throwexc::Bool, all::Bool, failfast::Bool)
   @ Base ./task.jl:513
 [2] _wait_multiple
   @ ./task.jl:417 [inlined]
 [3] waitany
   @ ./task.jl:397 [inlined]
 [4] prompt!(term::REPL.Terminals.TextTerminal, prompt::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
   @ REPL.LineEdit ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:2833
 [5] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
   @ REPL.LineEdit ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:2743
 [6] run_frontend(repl::REPL.LineEditREPL, backend::REPL.REPLBackendRef)
   @ REPL ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:1504
 [7] (::REPL.var"#75#81"{REPL.LineEditREPL, REPL.REPLBackendRef})()
   @ REPL ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:496
ERROR: TaskFailedException
Stacktrace:
  [1] try_yieldto(undo::typeof(ensure_rescheduled))
    @ Base ./task.jl:1100
  [2] wait()
    @ Base ./task.jl:1164
  [3] wait(c::GenericCondition{ReentrantLock}; first::Bool)
    @ Base ./condition.jl:139
  [4] wait
    @ ./condition.jl:134 [inlined]
  [5] take_buffered(c::Channel{Any})
    @ Base ./channels.jl:493
  [6] take!
    @ ./channels.jl:487 [inlined]
  [7] repl_backend_loop(backend::REPL.REPLBackend, get_module::Function)
    @ REPL ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:353
  [8] start_repl_backend(backend::REPL.REPLBackend, consumer::Any; get_module::Function)
    @ REPL ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:343
  [9] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_
```

Probably related to https://github.com/JuliaLang/julia/pull/54674 looking at the blame.

cc @IanButterworth

**Comment by IanButterworth** [comment]
Seems to be an issue with this line. Is there a more correct way to wait on input but not error on eof?
https://github.com/JuliaLang/julia/blob/ce38f08b166edfe33afa3129ae01ef349a67bfb0/stdlib/REPL/src/LineEdit.jl#L2834

**Comment by vtjnash** [comment]
Normally, the way to do that is to have 2 separate tasks, each responsible for one event, instead of having 3 tasks, each responsible for a semi-arbitrary set of events

**Comment by vtjnash** [comment]
Split each thing you want to do into a separate task, and don't try to complicatedly intermix them by ever using waitany

**Comment by IanButterworth** [comment]
I want to wait for the first of either
- new input being available from stdin, but not consume it, because it's consumed inside `match_input`
- a new function that needs executing from `s.async_channel`

The current design is the only way I know how to do that without blocking while waiting on stdin or throwing away new input on stdin

**Comment by vtjnash** [comment]
I think there are already a few other clients similar to that (e.g. see beep), though the current implementation of MVC does seem probably too tightly coupled between input and output to allow the clean injection of events that are not from the keymap

**Comment by vtjnash** [comment]
Reading from stdin without consuming the input introduces a data race into the design, which is probably why it crashes for some people periodically
